lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
-          // For partial pipeline compile, only fragment shader needs to adjust offset of root descriptor
-          // If there are more individual shader compile in future, we can add more stages here
-          userDataValue = DescRelocMagic | node->innerTable[0].set;
-        }
-        m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx++, userDataValue);
-        break;
+    unsigned *argIndex = &intfData->entryArgIdxs.resNodeValues[i];
+    unsigned userDataValue = node.offsetInDwords;
+    bool mustSpill = false;
+    switch (node.type) {
+
+    case ResourceNodeType::DescriptorTableVaPtr:
+      if (m_pipelineState->getShaderOptions(m_shaderStage).updateDescInElf &&
+          (m_shaderStage == ShaderStageFragment || m_shaderStage == ShaderStageVertex ||
+           m_shaderStage == ShaderStageCompute)) {
+        // Put set number to register first, will update offset after merge ELFs
+        // For partial pipeline compile, only fragment shader needs to adjust offset of root descriptor
+        // If there are more individual shader compile in future, we can add more stages here
+        userDataValue = DescRelocMagic | node.innerTable[0].set;
       }
+      break;
 
-      case ResourceNodeType::DescriptorBuffer:
-      case ResourceNodeType::DescriptorResource:
-      case ResourceNodeType::DescriptorSampler:
-      case ResourceNodeType::DescriptorTexelBuffer:
-      case ResourceNodeType::DescriptorFmask:
-        // Where a descriptor appears in the top-level table, it is accesssed
-        // via the spill table, rather than directly placed in sgprs.
-        assert(needSpill);
-        if (intfData->spillTable.offsetInDwords == InvalidValue)
-          intfData->spillTable.offsetInDwords = node->offsetInDwords;
-        m_pipelineState->getPalMetadata()->setUserDataSpillUsage(node->offsetInDwords);
-        break;
+    case ResourceNodeType::PushConst:
+    case ResourceNodeType::DescriptorBufferCompact:
+      break;
 
-      case ResourceNodeType::PushConst:
-      case ResourceNodeType::DescriptorBufferCompact: {
-        argTys.push_back(VectorType::get(Type::getInt32Ty(*m_context), node->sizeInDwords));
-        m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx, node->offsetInDwords,
-                                                            node->sizeInDwords);
-        userDataIdx += node->sizeInDwords;
-        break;
-      }
-      default: {
-        llvm_unreachable("Should never be called!");
-        break;
-      }
-      }
-    } else {
-      if (needSpill && intfData->spillTable.offsetInDwords == InvalidValue)
-        intfData->spillTable.offsetInDwords = node->offsetInDwords;
-      m_pipelineState->getPalMetadata()->setUserDataSpillUsage(node->offsetInDwords);
+    default:
+      // Any other descriptor type in the root table is accessed via the spill table.
+      mustSpill = true;
```

Comment:
why do need set mustSpill for all other resource types? I think we don't have this limitations.
as far as I know, some RPM shader put the resource/sampler in root.

Reply:
I think I added that limitation in an earlier change.

Because spir-v allows you to pass a pointer to an image or sampler descriptor through phi nodes and function calls, the earlier change properly separated the action of taking the pointer to the descriptor from the action of loading the descriptor.

With that separation, not spilling a root table image or sampler descriptor would require a peephole to link the two parts together. I did not implement that peephole because I thought that the only case of a root descriptor is a "dynamic" buffer descriptor.

What is an RPM shader? Some special internal shader not from the user program?

Reply:
RPM shaders are internal shaders, created in advance to execute fixed pipeline functions of graphics system. You can refer the shader sources and pipeline info in /pal/src/core/hw/gfxip/rpm/shaders from internal code base.